eth/protocols/snap/sync.go

```diff
@@ -0,0 +1,718 @@
+		incompletes := light.NewNodeSet()
+
+		for i, blob := range res.accounts {
+			// Decode the retrieved account
+			var acc state.Account
+			if err := rlp.DecodeBytes(blob, &acc); err != nil {
+				log.Error("Failed to decode full account", "err", err)
+				return err
+			}
+			// Retrieve any associated bytecode, if not yet downloaded
+			if !bytes.Equal(acc.CodeHash, emptyCode[:]) {
+				if code, err := s.db.Get(acc.CodeHash); err != nil || code == nil {
+					// Generate a random request ID (clash probability is insignificant)
+					id := uint64(rand.Int63())
+
+					// Track the request and send it to the peer
+					s.lock.Lock()
+					s.bytecodeReqs[peer.id] = &bytecodeRequest{
+						id:     id,
+						hashes: []common.Hash{common.BytesToHash(acc.CodeHash)},
+						cancel: cancel,
+					}
+					s.lock.Unlock()
+
+					if err := peer.RequestByteCodes(id, []common.Hash{common.BytesToHash(acc.CodeHash)}, maxRequestSize); err != nil {
+						return err
+					}
+					s.bytecodeRequests++
+
+					// Wait for the reply to arrive
+					res := <-s.bytecodeResps
+					s.bytecodeSynced += uint64(len(res.codes))
+
+					if len(res.codes) != 1 || res.codes[0] == nil {
+						return errors.New("protocol violation")
+					}
+					s.bytecodeBytes += common.StorageSize(len(res.codes[0]))
+					s.db.Put(acc.CodeHash, res.codes[0])
+				}
+			}
+			// Retrieve any associated storage trie, if not yet downloaded
+			if acc.Root != emptyRoot {
+				if node, err := s.db.Get(acc.Root[:]); err != nil || node == nil {
+					// Sync the contract's storage trie
+					complete, err := s.syncStorage(root, res.hashes[i], acc.Root, peer, cancel)
+					if err != nil {
+						return err
+					}
+					// If the storage sync is incomplete (missing boundary nodes
+					// across multiple requests), mark the account as incomplete
+					// to force self healing at the end,
+					if !complete {
+						if err := res.trie.Prove(res.hashes[i][:], 0, incompletes); err != nil {
+							panic(err) // Account range was already proven, what happened
+						}
+					}
+				}
+			}
+			// If the snapshot moved during contract sync, nuke out all remaining accounts
+			var interrupted bool
+			select {
+			case <-cancel:
+				for j := i + 1; j < len(res.hashes); j++ {
+					if err := res.trie.Prove(res.hashes[j][:], 0, incompletes); err != nil {
+						panic(err) // Account range was already proven, what happened
+					}
+				}
+				interrupted = true
+			default:
+			}
+			if interrupted {
+				// Sync was interrupted, restart next cycle at the current account,
+				// but leave next slot at wherever we were.
+				//
+				// TODO(karalabe): Special case account deletion in the next cycle or proof-lessness, musn't write
+				s.nextAcc = res.hashes[i]
+				break
+			}
+			// Account processed fully (may still be incomplete, but that's for
+			// trie node sync to complete), push the next account marker
+			s.nextAcc = common.BigToHash(new(big.Int).Add(res.hashes[i].Big(), big.NewInt(1)))
```

Comment:
This assumes that the accounts delivered are sorted. Is that guaranteed somewhere?

Reply:
I don't think it is. I added that as a testcase to the trie prover malicious-tests PR

Reply:
It must be sorted, that's a protocol requirement. Not sure if it's enforced anywhere though. If not, we should.